docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md

```diff
@@ -0,0 +1,274 @@
+snapshot data.
+
+## Use-cases
+
+What's the point of such read-only volumes?
+
+* **Restore snapshots selectively:** users may want to traverse snapshots,
+  restoring data to a writable volume more selectively instead of restoring
+  the whole snapshot.
+* **Volume backup:** users can't backup a live volume, they first need
+  to snapshot it. Once a snapshot is take, it still can't be backed-up,
+  as backup tools usually work with volumes (that are exposed as file-systems)
+  and not snapshots (which might have backend-specific format). What this means
+  is that in order to create a snapshot backup, users have to clone snapshot
+  data twice:
+
+  1. first time, when restoring the snapshot into a temporary volume from
+     where the data will be read,
+  1. and second time, when transferring that volume into some backup/archive
+     storage (e.g. object store).
+
+  The temporary backed-up volume will most likely be thrown away after the
+  backup transfer is finished. That's a lot of wasted work for what we
+  originally wanted to do! Having the ability to create volumes from
+  snapshots cheaply would be a big improvement for this use case.
+
+## Alternatives
+
+* _Snapshots are stored in `<subvolume>/.snap`. Users could simply visit this
+  directory by themselves._
+
+  `.snap` is CephFS-specific detail of how snapshots are exposed.
+  Users / tools may not be aware of this special directory, or it may not fit
+  their workflow. At the moment, the idiomatic way of accessing snapshot
+  contents in CSI drivers is by creating a new volume and populating it
+  with snapshot.
+
+## Design
+
+Key points:
+
+* Volume source is a snapshot, volume access mode is `*_READER_ONLY`.
+* No actual new subvolumes are created in CephFS.
+* The resulting volume is a reference to the source subvolume snapshot.
+  This reference would be demarked in `Volume.volume_context` map.
+* Mounting such volume means mounting the respective CephFS subvolume
+  and exposing the snapshot to workloads.
+* Let's call a *shallow read-only volume with a subvolume snapshot
+  as its data source* just a *shallow volume* from here on out for brevity.
+
+### Controller operations
+
+Care must be taken when handling life-times of relevant storage resources.
+When a shallow volume is created, what would happen if:
+
+* _Parent subvolume of the snapshot is removed while the shallow volume
+  still exists?_
+
+  This shouldn't be a problem already. The parent volume has either
+  `snapshot-retention` subvol feature in which case its snapshots remain
+  available, or if it doesn't have that feature, it will fail to be deleted
+  because it still has snapshots associated to it.
+* _Source snapshot from which the shallow volume originates is removed while
+  that shallow volume still exists?_
+
+  We need to make sure this doesn't happen and some book-keeping
+  is necessary. Ideally we could employ some kind of reference counting.
+
+#### Reference counting for shallow volumes
```

Comment:
Do we need to consider PVC-PVC clone case from PVC creating as shallow volume? to allow disallow the PVC-PVC clone operation?

Reply:
Good catch!

* request to create **shallow volume** with **shallow volume** as its source: basically a new reference to the parent snapshot,
* request to create **regular volume** with **shallow volume** as its source: equivalent for a request to create a regular volume with snapshot as its source,
* request to create **shallow volume** with **regular volume** as its source: this doesn't make sense and should exit with an error.

Reply:
https://github.com/ceph/ceph-csi/pull/2148/commits/4db796a7decf17050bfd66fb3ed9c54ebe0acb68